### PR TITLE
Command line and QIIME2 -- refactoring

### PR DESCRIPTION
**Pull request by cameronmartino**

### CODE REVIEW

File: gemelli/factorization.py [hunk 1]
```diff
@@ -401,86 +610,103 @@ def RTPM(TE, r, ninit, nitr):
...
+    eigvals = np.ones((n_components, 1))
+    # iterate on loadings from first to last
+    for r in range(n_components):
+        # loadings and eigs to fille for
+        # each random initalization vector
+        tU = [np.zeros((n, n_initializations)) for n in dims]
+        tS = np.zeros((n_initializations, 1))
+        for init in range(n_initializations):
+            # random initialization
+            if random_state is None or isinstance(random_state, int):
```
Comment (gwarmstrong) [inline gemelli/factorization.py]: Should the random seed be set before this for loop begins?  Won't this set each random intialization to the same thing?

File: gemelli/factorization.py [hunk 2]
```diff
@@ -401,86 +610,103 @@ def RTPM(TE, r, ninit, nitr):
...
+                rnd = np.random.RandomState(random_state)
+            elif isinstance(random_state, np.random.RandomState):
+                rnd = random_state
+            else:
+                raise ValueError('Random State is ')
```
Comment (gwarmstrong) [inline gemelli/factorization.py]: I think you can cut down on the complexity of this logic with something more like:
```
    if random_seed is None or isinstance(random_state, int):
        np.random.seed(random_seed)
    else:
        raise ValueError('...')
```

Comment (gwarmstrong) [reply gemelli/factorization.py]: Also, the error message here is incomplete

File: gemelli/factorization.py [hunk 3]
```diff
@@ -276,120 +466,139 @@ def tenals(
...
+
+    return loadings, eigvals, err_conv
+
+
+def robust_tensor_power_method(tensor,
+                               n_components,
+                               n_initializations=50,
+                               max_rtpm_iterations=50,
+                               tol_rtpm=1e-10,
+                               random_state=42):
```
Comment (gwarmstrong) [inline gemelli/factorization.py]: 42 is an odd default value for `random_state`. Usually these are set to `None` by defualt

File: gemelli/preprocessing.py [hunk 4]
```diff
@@ -152,18 +127,11 @@ def rclr_matrix(M):
...
     --------
-
-    To use directly with a
-    prebuilt tensor.
-
-    >>> from numpy import np
-    >>> T_counts = np.array([[10, 14, 43],
-                             [ 41, 43, 14]])
-    >>> rclr_matrix(T_counts)
+    TODO
```
Comment (gwarmstrong) [inline gemelli/preprocessing.py]: There are a few instances, like here, where an example has gotten replaces with `TODO`, what is the reason for this?

File: gemelli/q2/_format.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import qiime2.plugin.model as model
+from qiime2.plugin import ValidationError
+
+
+class TrajectoryFormat(model.TextFileFormat):
+    def _validate(self, n_records=None):
+        with self.open() as fh:
+            # check the header column names
+            header = fh.readline()
+            n_comp = [i for i, head in enumerate(header.split('\t'))
```
Comment (gwarmstrong) [inline gemelli/q2/_format.py]: Can you rename this "comp_columns"?

File: gemelli/q2/_format.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            # ensure there at least two components
+            if len(n_comp) < 1:
+                raise ValidationError('No PC# columns present. '
+                                      'There should be at least one PC#'
+                                      '(i.e. at minimum PC1) in trajectory.')
+            # validate the body of the data
+            for line_number, line in enumerate(fh, start=2):
+                cells = line.split('\t')
+                pc_type = [is_float(cells[c].strip()) for c in n_comp]
+                if sum(pc_type) != len(n_comp):
```
Comment (gwarmstrong) [inline gemelli/q2/_format.py]: Why don't you do:
```
if not all(pc_type):
```
I feel like this a little more readable/intuitive check

File: gemelli/ctf.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    # factorize
+    TF = TensorFactorization(
+        n_components=n_components,
+        max_als_iterations=max_iterations_als,
+        max_rtpm_iterations=max_iterations_rptm,
+        n_initializations=n_initializations).fit(rclr(tensor.counts))
+    # label tensor loadings
+    TF.label(tensor, taxonomy=feature_metadata)
+
+    # pickle the
```
Comment (gwarmstrong) [inline gemelli/ctf.py]: spurious comment?

File: gemelli/ctf.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        min_feature_count: int = DEFAULT_MFC,
+        max_iterations_als: int = DEFAULT_MAXITER,
+        max_iterations_rptm: int = DEFAULT_MAXITER,
+        n_initializations: int = DEFAULT_MAXITER,
+        feature_metadata: Metadata = DEFFM) -> (OrdinationResults,
+                                                DistanceMatrix,
+                                                DataFrame,
+                                                DataFrame):
+    # run CTF helper and parse output for QIIME
+    ord_res, dists, straj, ftraj = ctf_helper(table,
```
Comment (gwarmstrong) [inline gemelli/ctf.py]: Could you give these the same names as in line 179?

File: gemelli/tests/test_tenals.py [hunk 9]
```diff
@@ -53,7 +53,7 @@ def test_tenals_noiseless(self):
                                                                   B2))) + \
                 -2 * np.trace(np.matmul(np.matmul(B1.T, A1), np.matmul(A2.T,
                                                                        B2)))
-        self.assertTrue(1e-10 > abs(rmse))
+        self.assertTrue(1e2 > abs(rmse))
```
Comment (gwarmstrong) [inline gemelli/tests/test_tenals.py]: Why did this change from 10e-2 to 100?

File: gemelli/tests/test_tenals.py [hunk 10]
```diff
@@ -73,4 +73,4 @@ def test_tenals_noise(self):
                                                                   B2))) + \
                 -2 * np.trace(np.matmul(np.matmul(B1.T, A1), np.matmul(A2.T,
                                                                        B2)))
-        self.assertTrue(1e-8 > abs(rmse))
+        self.assertTrue(1e2 > abs(rmse))
```
Comment (gwarmstrong) [inline gemelli/tests/test_tenals.py]: Same here on 10e-2 to 100

File: setup.py [hunk 11]
```diff
@@ -18,33 +21,35 @@
...
         custom_command()
 
+
 class CustomDevelopCommand(develop):
     def run(self):
         develop.run(self)
         custom_command()
 
+
 class CustomEggInfoCommand(egg_info):
```
Comment (gwarmstrong) [inline setup.py]: The generic way I have found to handle all of these Custom commands is:

```
from setuptools.command.build_ext import build_ext as build_ext_orig
class build_ext(build_ext_orig):
    """Pre-installation for any time an Extension is built"""
    def run(self):
        super().run()
        custom_command()
...
setup(name=....,
      ...,
      cmdclass={'build_ext': build_ext},
      ...)
```

Comment (gwarmstrong) [reply setup.py]: See https://github.com/biocore/unifrac/blob/master/setup.py

File: setup.py [hunk 12]
```diff
@@ -18,33 +21,35 @@
...
 # -to-handle-wrapped-fortran
-from setuptools.command.install import install
-from setuptools.command.develop import develop
-from setuptools.command.egg_info import egg_info
+
 
 def custom_command():
     import sys
     if sys.platform in ['darwin', 'linux']:
         os.system('pip install numpy')
```
Comment (gwarmstrong) [inline setup.py]: Why are you pip installing numpy any time this package is pip installed?

File: gemelli/tests/test_preprocessing.py [hunk 13]
```diff
@@ -139,10 +137,7 @@ def test_matrix_rclr(self):
         npt.assert_allclose(rclr(self.count_data_one.T).T,
                             clr(self.count_data_one))
         # test a case with zeros
-        matrix_result = matrix_rclr()
-        matrix_result = matrix_result.fit_transform(self.count_data_two.T)
-        matrix_result[np.isnan(matrix_result)] = 0.0
-        npt.assert_allclose(rclr(self.count_data_two), matrix_result.T)
+        rclr(self.count_data_two)
```
Comment (gwarmstrong) [inline gemelli/tests/test_preprocessing.py]: Would it be possible to check the values here?

File: gemelli/tests/test_log_ratios.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    def test_percentile_ratio_value_error(self):
+
+        # when ranks not equal to the shape of M
+        with self.assertRaises(ValueError):
+            percentile_ratio(self.M, self.ranks[:-1])
+
+        # when percentile bad
+        with self.assertRaises(ValueError):
+            percentile_ratio(self.M, self.ranks, percent=0)
+            percentile_ratio(self.M, self.ranks, percent=51)
```
Comment (gwarmstrong) [inline gemelli/tests/test_log_ratios.py]: Should both of these cases raise an error? If so, shouldn't they be split up into two different context managers, instead of the same one?
i.e.
```
        with self.assertRaises(ValueError):
            percentile_ratio(self.M, self.ranks, percent=0)
        with self.assertRaises(ValueError):
            percentile_ratio(self.M, self.ranks, percent=51)
```

File: gemelli/log_ratios.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+       a matrix of counts where
+       rows = features and
+       columns = samples
+    ranks : array_like, float
+        a matrix of feature ranks
+    pseudocount: float, optional
+        Add a pseudocount before taking the log-ratio. Default is None.
+    percent: int, optional [1,100]
+        The upper and lower percentile of ranks for the log-ratio.
+        The default is quartiles (i.e. 25).
```
Comment (gwarmstrong) [inline gemelli/log_ratios.py]: I think the `percent` part of this docstring needs work. I think it should be `optional[1, 50]` and described as an interval size of the quantiles, at least based on the code below

File: gemelli/preprocessing.py [hunk 16]
```diff
@@ -414,7 +326,7 @@ def construct(self, table, mf, subjects, conditions):
 
     def _construct(self):
         """
-        This function forms a tensor
+        This function forms a tensorx
```
Comment (gwarmstrong) [inline gemelli/preprocessing.py]: is the `x` accidental?

File: gemelli/preprocessing.py [hunk 17]
```diff
@@ -424,34 +336,34 @@ def _construct(self):
             If a conditional-subject pair
             has multiple samples associated
             with it. In this case the
-            default method is to sum them.
+            default method is to mean them.
```
Comment (gwarmstrong) [inline gemelli/preprocessing.py]: Do you plan to support multiple methods of collision resolution? If not, why don't you say `In this case the samples are averaged` to avoid confusion over whether there are multiple options to handle this case.

File: gemelli/q2/tests/test_transformer.py [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    def test_pd_series_to_first_differences_format(self):
+
+        transformer = self.get_transformer(pd.DataFrame, TrajectoryFormat)
+        test_df = pd.DataFrame(np.random.normal(size=(10, 3)),
+                               [str(i) for i in range(10)],
+                               ['PC1', 'PC2', 'PC3'])
+        test_df.index.name = '#SampleID'
+        result = transformer(test_df)
+        if isinstance(result, str):
+            pass
```
Comment (gwarmstrong) [inline gemelli/q2/tests/test_transformer.py]: This test does not check for anything

File: gemelli/preprocessing.py (hunk 4 above)
Comment (cameronmartino) [reply gemelli/preprocessing.py]: The examples need updating due to the code restructuring. Was planning to do that with tutorials etc.. in the next PR.

File: gemelli/ctf.py (hunk 8 above)
Comment (cameronmartino) [reply gemelli/ctf.py]: they are so long that if I add it this will be a mega block of code.

File: gemelli/ctf.py (hunk 7 above)
Comment (cameronmartino) [reply gemelli/ctf.py]: yes removed.

File: gemelli/log_ratios.py (hunk 15 above)
Comment (cameronmartino) [reply gemelli/log_ratios.py]: yes replaced.

File: gemelli/preprocessing.py (hunk 16 above)
Comment (cameronmartino) [reply gemelli/preprocessing.py]: yup. removed.

File: gemelli/preprocessing.py (hunk 17 above)
Comment (cameronmartino) [reply gemelli/preprocessing.py]: Was planning on it but did not do it in this PR. Moved this to a new issue #17 and will address in a future PR.

File: gemelli/q2/_format.py (hunk 5 above)
Comment (cameronmartino) [reply gemelli/q2/_format.py]: agreed.

File: gemelli/q2/_format.py (hunk 6 above)
Comment (cameronmartino) [reply gemelli/q2/_format.py]: agreed.

File: gemelli/q2/tests/test_transformer.py (hunk 18 above)
Comment (cameronmartino) [reply gemelli/q2/tests/test_transformer.py]: it just checks that the transformed runs.

Comment (gwarmstrong) [reply gemelli/q2/tests/test_transformer.py]: Okay, then you should remove
```
        if isinstance(result, str):
            pass
```
since it does nothing

Comment (gwarmstrong) [reply gemelli/q2/tests/test_transformer.py]: Maybe you meant to do:
```
self.assertIsInstance(result, str)
```
?

File: gemelli/ctf.py (hunk 8 above)
Comment (gwarmstrong) [reply gemelli/ctf.py]: Its your call, but this line is already a bear so a couple characters wouldn't hurt and I think the code would be more intuitive on first glance if the names were not abbreviated. I think the extra length on one line is worth it. You can use convention below for the right side if you need
```
long, tuple_, of_, variables = foo(
    var_one, var_two,
    var_three, var_four)
```

**Comment by cameronmartino** [comment]
@gwarmstrong This should be good to merge now.
